Database Implementation

sharick123 opened the pull request:
Set up a MySQL database which stores username and password for registration. The players can login using the same details. 
Incorrect username/password combinations return clients to the game homepage and displays error in the console. There is an optional system alert option within comments to display the same. 

Note: This database is set up locally on the client. Remember to use your own host, user and password in server.js file.

esefton1 committed:
Create database.js

creating a file for the database

esefton1 committed:
Update database.js

start of database creation

esefton1 committed:
Update database.js

fixed missing bracket

esefton1 committed:
Update database.js

Commit:
Register database set up

esefton1 committed:
Update database.js

changed from querying database to createing database

Commit:
Registeration database setup

Commit:
Login Working. Needs better prompts for handling errors.

Commit:
Removing server files used for testing.

Commit:
Added an optional System Alert for incorrect username/password.

esefton1 reviewed (changes requested):
Looks good but a couple of syntax issues need addressed so will make some small changes

esefton1 committed:
Update server.js

changing some syntax

esefton1 committed:
Update server.js

More syntax changes

esefton1 committed:
Update server.js

Syntax changes

esefton1 committed:
Update server.js

Syntax changes

esefton1 committed:
Update server.js

changing some syntax

esefton1 committed:
Update server.js

syntax changes

kymckay commented on package.json:
```diff
@@ -37,6 +37,11 @@
+    "express": "^4.16.4",
+    "express-validator": "^6.6.1",
+    "mysql": "^2.18.1",
+    "nodemon": "^2.0.6"
```
As `nodemon` is not needed for deployment it shouldn't be listed in the dependencies, it could be made into a dev dependency, but currently our `package.json` is not setup to use it. Should probably remove it from here (you can still install it locally for your own development use).

```suggestion
```

kymckay commented on server/server.js:
```diff
@@ -1,25 +1,110 @@
+
 const app = express();
-const port = 8080;
+const port = 8070;
```
```suggestion
const port = 8080;
```

No need to change this now

kymckay commented on package.json:
```diff
@@ -37,6 +37,11 @@
+    "alert": "^5.0.5",
+    "body-parser": "^1.19.0",
+    "express": "^4.16.4",
+    "express-validator": "^6.6.1",
```
Looks like this package is not used in the live build and so should not be listed here

```suggestion
```

kymckay commented on package.json:
```diff
@@ -37,6 +37,11 @@
   },
   "dependencies": {
-    "express": "^4.16.4"
+    "alert": "^5.0.5",
```
Looks like this package is not used in the live build and so should not be listed here (this is more significant than it sounds)
```suggestion
```

kymckay commented on server/server.js:
```diff
@@ -1,25 +1,110 @@
+
+// const { check } = require('express-validator');
+
+app.use(bodyParser.urlencoded({ extended: false }));
```
Should probably add a comment to explain the purpose of this line (I'm unsure what it's doing)

kymckay commented on server/server.js:
```diff
@@ -1,25 +1,110 @@
+const con = mysql.createConnection({
+  host: 'localhost', // use your own hostname
+  user: 'root', // use your mysql username
+  password: 'Aceg0864', // use your password
```
I don't think this password should be in the remote repository, there are probably standard ways to best handle this we can look into (I've not done it before, but came across similar situations). I hope this isn't a password you use elsewhere as it's now publicly compromised.

kymckay commented on server/server.js:
```diff
@@ -1,25 +1,110 @@
 app.post('/register', (req, res) => {
-    // TODO Register new user, check if already exists, handle client side after register
-})
+  // TODO Register new user, check if already exists, handle client side after register
```
```suggestion
```

kymckay commented on server/server.js:
```diff
@@ -1,25 +1,110 @@
+});
+
+app.post('/login', (req, res) => {
+  // TODO Login existing user if exists, refuse otherwise
```
```suggestion
```

kymckay commented on server/server.js:
```diff
@@ -1,25 +1,110 @@
+
+  con.connect(function (err) {
+    con.query(credentials, [name, pass], (err, data, fields) => {
+      if (err) {
+        console.log("Failed to login: " + err);
+      } else if (data.length > 0) {
+        console.log('Successfully logged in.');
+        res.redirect('play.html');
+      } else {
+        console.log('Incorrect username or password.');
+        // alert('Incorrect username or password.');
+        // res.sendStatus(500);
+        // return;
```
These seem fine for now, we should look into making it more response on the client-side (respond with the messages and show them in the DOM somehow). May need to change how the actual form requests are working to achieve that.

kymckay commented on server/server.js:
```diff
@@ -1,25 +1,110 @@
+  // TODO Register new user, check if already exists, handle client side after register
+
+  const name = req.body.user;
+  const pass = req.body.pword;
```
Should probably open an issue to look into how to store passwords securely (not just as a string). That way if the database is ever compromised the passwords are not exposed.

kymckay replied:
Think the group from last year had code using a package to do something like this.

kymckay commented on server/server.js:
```diff
@@ -1,25 +1,110 @@
+  const query2 = 'INSERT INTO players (username, password) VALUES (?, ?)';
+
+  con.connect(function (err) {
+    con.query(query2, [name, pass], (err, results, fields) => {
```
Ah nice, I wonder if this function is sanitising the user input to prevent SQL injection. Should look into that and document/comment it if so - so that it's clear.

kymckay commented on server/server.js:
```diff
@@ -1,25 +1,110 @@
+        return;
+      }
+      console.log('Registered new user successfully.');
+      res.redirect('index.html');
```
See what you mean about the redirect, think it's fine for now, but like below may want to investigate making it more dynamic/responsive immediately on the same page.

kymckay commented on server/server.js:
```diff
@@ -1,25 +1,110 @@
+  if (err) throw err;
+  console.log('Connected!');
+
+  con.query('DROP DATABASE IF EXISTS steak', function (err, result) {
```
Is this dropping the database every time the server is created? If it is, I'm guessing that was for development purposes, should probably not happen in the live code.

kymckay reviewed (changes requested):
Initial review, seems like a solid foundation for the database setup and login. Just needs some small clean up and then further iterative improvement once this PR is merged 👍

kymckay committed:
Merge branch 'master' into database-demo

kymckay commented:
I went ahead and fixed the merge conflicts here, took the opportunity to resolve some of my review comments that made sense for a merge.

May investigate factoring the database code out into its own file if I have time tonight (for readability/encapsulation).

kymckay committed:
Factor database code out to new file

Just improving encapsulation and readability

kymckay committed:
Add comment explaining body parser's use

kymckay committed:
Read SQL details in from separate file

Removes the risk of passwords being added to the git history

kymckay commented:
Went ahead and solved my outstanding comments. The only thing I think necessary before merge now is to add possibly to run the server without the database just for testing (saves everyone installing MySQL).

I'll open up two issues:
- Storing the passwords securely
- Making the index page update based on logged-in status/request failure/etc. (I have some idea on how we could do this).

kymckay committed:
Merge master and implement database responses

kymckay commented:
Okay, I've now got this sending out appropriate AJAX responses to the clients based on the actual database queries.

All that's left to do is set it up for deployment on Heroku and add in some failsafe for if a database connection can't be established (just run the server anyway and don't store anything).

kymckay committed:
Prepare database config for Heroku deployment

kymckay committed:
Validate username and password more accurately

Impose length limits as per the database datatypes

kymckay committed:
Prepare database to store statistics

kymckay committed:
Fix my validation regex

Also improve bad username message

kymckay committed:
Use bodyparser middleware only when needed

Also took the oppertunity to make the variable names consistent

kymckay committed:
Fix user registration

- Forgot to initalise the statistic values
- In my testing realised we want to log any errors on the server for
  diagnosis
- Don't want to throw, so server keeps going

kymckay committed:
Better handle DB error messages for client

By setting the response status to 500 the message neevr actually gets
displayed. It probably makes more sense this way anyway since response
500 should occur if the server itself messes up.

kymckay committed:
Fix content type of login requests

kymckay committed:
Remove unnecessary console statement

kymckay committed:
Allow use of any database name desired

Mostly useful for the ClearDB Heroku setup as we cannot control the DB
name

kymckay committed:
Fix Heroku ClearDB connection
